### PR TITLE
Elide dimensions

### PR DESCRIPTION
**Pull request by jkusa**
#625  Description

Plumbing work to allow for elide dimension type ahead features

## Proposed Changes

- Add insert method to keg
- Add metadata fields to elide dimensions
- Metadata serializers to provide models

## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: packages/data/addon/services/keg.ts [hunk 1]
```diff
@@ -85,6 +91,51 @@ export default class KegService extends Service {
     return this.pushMany(type, [rawRecord], options).firstObject;
   }
 
+  /**
+   * Inserts a single record into the store
```
Comment (kevinhinterlong) [inline packages/data/addon/services/keg.ts:95]: Maybe we can update the `pushMany` method to `create` and then call `insertMany` to make the difference more clear

File: packages/data/addon/services/keg.ts [hunk 2]
```diff
@@ -85,6 +91,51 @@ export default class KegService extends Service {
...
+   *
+   * @param type - type name of the model type
+   * @param record - record to be inserted into the keg
+   * @param options - config object
+   */
+  insertMany(type: string, records: Array<KegRecord>, options: Partial<InsertOptions> = {}): EmberArray<KegRecord> {
+    const recordKeg = this._getRecordKegForType(type);
+    const idIndex = this._getIdIndexForType(type);
+    const namespace = options.namespace || this.defaultNamespace;
+    const identifierField = options?.identifierField || KegService.identifierField;
```
Comment (mattgraham1995) [inline packages/data/addon/services/keg.ts:116]:
```suggestion
    const identifierField = options.identifierField || KegService.identifierField;
```

File: packages/data/addon/services/keg.ts [hunk 3]
```diff
@@ -85,6 +91,51 @@ export default class KegService extends Service {
...
+  }
+
+  /**
+   * Inserts an array of records into the store
+   *
+   * @param type - type name of the model type
+   * @param record - record to be inserted into the keg
+   * @param options - config object
+   */
+  insertMany(type: string, records: Array<KegRecord>, options: Partial<InsertOptions> = {}): EmberArray<KegRecord> {
```
Comment (mattgraham1995) [inline packages/data/addon/services/keg.ts:112]: Is pushMany being deprecated or is there a significant difference between this and that?

**Review by kevinhinterlong — APPROVED** [review]

File: packages/data/tests/unit/services/keg-test.js [hunk 4]
```diff
@@ -113,13 +105,19 @@ module('Unit | Service | keg', function(hooks) {
...
+    assert.equal(Record1, foundRecord, 'after inserting a record it can be found');
+  });
+
+  test('pushing a record with an existing id update the record in the keg', function(assert) {
+    //Insert initial record
+    const pushedRecord = Keg.push('record', RawRecord1);
 
-    //Push new record with same id
+    //Insert new record with same id
     Keg.push('record', { id: 1, description: 'updated' });
```
Comment (mattgraham1995) [inline packages/data/tests/unit/services/keg-test.js:121]: Should this be using `insert` or `push`?

File: packages/data/tests/unit/services/keg-test.js [hunk 5]
```diff
@@ -139,26 +137,39 @@ module('Unit | Service | keg', function(hooks) {
...
+    const newRecord = RecordFactory.create({ ...RawRecord1, description: 'updated' });
+    Keg.insert('record', newRecord);
+
+    let fetchedRecord = Keg.getById('record', 1);
 
-    let pushedRecord = Keg.push('notRecord', Record1, {
+    assert.deepEqual(
+      Keg.all('record'),
+      [newRecord],
+      'Pushing a record into the keg with an existing id does not add a new record'
```
Comment (mattgraham1995) [inline packages/data/tests/unit/services/keg-test.js:153]:
```suggestion
      'Inserting a record into the keg with an existing id does not add a new record'
```

**Review by mattgraham1995 — COMMENTED** [review]
`insert` vs `push` seems a little confusing

File: packages/data/addon/services/keg.ts (hunk 2 above)
Comment (jkusa) [reply packages/data/addon/services/keg.ts:116]: `options` is defaulted to `{}`

File: packages/data/addon/services/keg.ts (hunk 3 above)
Comment (jkusa) [reply packages/data/addon/services/keg.ts:112]: `insert` does not populate a model (`model.create(payload)`, that is the main diff. But we will deprecated push methods at some point. Maybe when we remove keg as a service?

File: packages/data/addon/services/keg.ts (hunk 2 above)
Comment (mattgraham1995) [reply packages/data/addon/services/keg.ts:116]: right that's why the optional chaining isn't needed, right? Or if it is needed, it should be added to the line above where we access `options.namespace`

Comment (jkusa) [reply packages/data/addon/services/keg.ts:116]: ah, read your comment the other way.

File: packages/data/tests/unit/services/keg-test.js [hunk 6]
```diff
@@ -113,13 +105,19 @@ module('Unit | Service | keg', function(hooks) {
...
+    const foundRecord = Keg.getById('record', 1, 'navi');
+    assert.equal(Record1, foundRecord, 'after inserting a record it can be found');
+  });
+
+  test('pushing a record with an existing id update the record in the keg', function(assert) {
+    //Insert initial record
+    const pushedRecord = Keg.push('record', RawRecord1);
 
-    //Push new record with same id
+    //Insert new record with same id
```
Comment (jkusa) [inline packages/data/tests/unit/services/keg-test.js:120]:
```suggestion
    //Push new record with same id
```

**Review by mattgraham1995 — APPROVED** [review]

**Pull request merged**